private/buf/buffetch/ref_parser.go

```diff
@@ -173,23 +180,28 @@ func newSourceOrModuleRefParser(logger *zap.Logger) *refParser {
 			internal.WithDirFormat(formatDir),
 			internal.WithModuleFormat(formatMod),
 		),
+		tracer: otel.GetTracerProvider().Tracer("bufbuild/buf"),
 	}
 }
 
 func (a *refParser) GetRef(
 	ctx context.Context,
 	value string,
 ) (Ref, error) {
-	ctx, span := trace.StartSpan(ctx, "get_ref")
+	ctx, span := a.tracer.Start(ctx, "get_ref")
 	defer span.End()
 	parsedRef, err := a.getParsedRef(ctx, value, allFormats)
 	if err != nil {
+		span.RecordError(err)
```

Comment:
I know that we tend to dislike named return types but in all of these instances it'd make everything a lot easier:
```go

func (a *refParser) GetRef(
	ctx context.Context,
	value string,
) (ref Ref, err error) {
	defer func(){ 
		if err != nil {
			span.RecordError(err)
			span.SetStatus(codes.Error, err.Error())
		}
	}
	ctx, span := a.tracer.Start(ctx, "get_ref")
	defer span.End()
	parsedRef, err := a.getParsedRef(ctx, value, allFormats)
```

Also means that when the function grows there isn't the risk of someone forgetting to record the error

Reply:
Another idea would be to use generics and whatever `func (ctx, string) (T, error)` function gets passed in gets traced with the name of the function:
```go
func TraceFunc[T any](tracer trace.Tracer, f func(ctx context.Context, value string) (T, error), ctx context.Context, value string) (T, error) {
	funcName := runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
	ctx, span := tracer.Start(ctx, funcName)
	retVal, err := f(ctx, value)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
	}
	return retVal, err
}
```

Reply:
This is wayyy outside the scope of this PR lol, just bike shedding :)

Reply:
Yeah, this one is kind of difficult for me, because I didn't want to heavily warp the patterns and texture of the code base around the tracing instrumentation. There needs to be a decision made here around the way we want to do debug logs before we commit to larger changes. The debug log instrumentation patterns are also a little bit inconsistent, so it's also important to figure out the rules.

I'll take a look at places where it might greatly benefit from the `defer` error handler, but otherwise, I am inclined to try and keep it as consistent as possible with existing patterns. I do agree that it can be a bit brittle, especially for long functions, so I think that's where I'd like to start with the error handler.